Sources/SwiftParser/Parameters.swift

```diff
@@ -0,0 +1,298 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_spi(RawSyntax) import SwiftSyntax
+
+// MARK: - Traits
+
+protocol RawParameterTrait: RawSyntaxNodeProtocol {
+  var trailingComma: RawTokenSyntax? { get }
+}
+
+extension RawFunctionParameterSyntax: RawParameterTrait {}
+extension RawClosureParameterSyntax: RawParameterTrait {}
+extension RawEnumCaseParameterSyntax: RawParameterTrait {}
+
+protocol RawParameterListTrait: RawSyntaxNodeProtocol {
+  associatedtype ParameterSyntax: RawParameterTrait
+
+  init(elements: [ParameterSyntax], arena: __shared SyntaxArena)
+}
+
+extension RawFunctionParameterListSyntax: RawParameterListTrait {}
+extension RawClosureParameterListSyntax: RawParameterListTrait {}
+extension RawEnumCaseParameterListSyntax: RawParameterListTrait {}
+
+protocol RawParameterClauseTrait: RawSyntaxNodeProtocol {
+  associatedtype ParameterList: RawParameterListTrait
+
+  init(
+    _ unexpectedBeforeLeftParen: RawUnexpectedNodesSyntax?,
+    leftParen: RawTokenSyntax,
+    _ unexpectedBetweenLeftParenAndParameterList: RawUnexpectedNodesSyntax?,
+    parameterList: ParameterList,
+    _ unexpectedBetweenParameterListAndRightParen: RawUnexpectedNodesSyntax?,
+    rightParen: RawTokenSyntax,
+    _ unexpectedAfterRightParen: RawUnexpectedNodesSyntax?,
+    arena: __shared SyntaxArena
+  )
+}
+
+extension RawParameterClauseSyntax: RawParameterClauseTrait {}
+extension RawClosureParameterClauseSyntax: RawParameterClauseTrait {}
+extension RawEnumCaseParameterClauseSyntax: RawParameterClauseTrait {}
+
+// MARK: Parameter parsing
+
+extension Parser {
+  mutating func parseFunctionParameter() -> RawFunctionParameterSyntax {
```

Comment:
There's a fair bit of duplication between these. Did you find that made more sense then having options on a `parseParameterLikeDecl(...)` sort of function? `RawParameterTrait` could include an `init` that just has alllll the things, where we ignore params that don't make sense where necessary.

Reply:
I wanted to get the compile-time safety of not forgetting to pass one of the required arguments and I think we can only reasonably to do was to duplicate it.

Reply:
You can still have separate `parseFunctionParameter` etc. for callers outside. It'd just be `fileprivate` within `Parameters`.